src/tree/ResourceCache.ts

```diff
@@ -0,0 +1,120 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { ResourceGroup } from "@azure/arm-resources";
+import { IActionContext, nonNullProp } from "@microsoft/vscode-azext-utils";
+import { AppResource } from "@microsoft/vscode-azext-utils/hostapi";
+import { ThemeIcon } from "vscode";
+import { GroupBySettings } from "../commands/explorer/groupBy";
+import { createAzureExtensionsGroupConfig } from "../utils/azureUtils";
+import { localize } from "../utils/localize";
+import { settingUtils } from "../utils/settingUtils";
+import { AppResourceTreeItem } from "./AppResourceTreeItem";
+import { GroupTreeItemBase } from "./GroupTreeItemBase";
+import { ResourceGroupTreeItem } from "./ResourceGroupTreeItem";
+import { SubscriptionTreeItem } from "./SubscriptionTreeItem";
+
+export type GroupTreeMap = { [groupTreeId: string]: GroupTreeItemBase };
+export class ResourceCache {
+    private _subscriptionTreeItem: SubscriptionTreeItem;
+    private _resourceGroups: ResourceGroup[];
+    private _appResources: AppResourceTreeItem[];
+
+    public constructor(sub: SubscriptionTreeItem) {
+        this._subscriptionTreeItem = sub;
+        this.resetCache();
+    }
+
+    private resetCache(): void {
+        this._appResources = [];
+        this._resourceGroups = [];
+    }
+
+    public get appResources(): AppResourceTreeItem[] {
+        return this._appResources;
+    }
+
+    public set appResources(resources: AppResource[]) {
+        const addedResources = this.resolveProvidedResources(this._appResources, resources);
```

Comment:
I don't feel strongly about it, but it might be good to make `resolveProvidedResources` something like `diffProvidedResources`, and have it return a tuple of added and removed. And then, handle both the adding *and* the removing here, rather than adding here, and removing within `resolveProvidedResources`.

Reply:
I originally  was going to do that, but having this duplicated kind of bothered me:
```
for (const deleted of deletedResources) {
            const index = cachedResources.findIndex(cachedResource => cachedResource.id === deleted.id);
            cachedResources.splice(index, 1);
        }
```

I could always just make that a util function as well, I suppose.

Reply:
I like that idea